### DISCUSSION TITLE
What shall we do with unsupported?

### DISCUSSION THREAD

**Discussion by osmihran** [gismo/gismo#405] [Q&A] [answered]
Hello everyone!

This is a question we touched during the developer days. Since it is a long-term decision, this might be a better platform to discuss it than slack.

It is my understanding that there is a lot of good functionality in `unsupported` but it is rotting, i.e., not all of it compiles and works anymore. That is certainly unfortunate. How do we envision its future? I can imagine several options and there certainly have already been discussions in this direction initiated by @mmoelle1, so feel free to correct me.

### Option A
@filiatra suggested to break `unsupported` into smaller chunks and turn these into modules, much as what @Shamanskiy did with `gsElasticity`.
#### Pros
+ The big task ("fix unsupported") gets divided into more manageable tasks ("take care of this module").
+ The modules can be administered by various people, thus dividing the burden.
+ The user of the library has to compile only the parts he or she needs.
#### Cons
+ Quite some administrative gets repeated for each module (CMake files, CI frameworks, unit testing, installation instructions etc.).
+ In a way, the complexity also increases: the packages may depend on each other and there are more possible combinations of git commits and releases interacting with each other.
+ It is not clear (at least to me), how changes in `stable` are to propagate into the other modules. What happens when the signature of a function in `stable` changes? The person doing such a change in `stable` should probably provide the corresponding fix in all packages as well but is this realistic? Do they have the means to notice it and actually do it? A maintainer of a module seems to have no way of saying "our module works with G+Smo, commit XYZ, do not update to the newest version of `stable` until we adapt our module."

In any case, if we decide to go this way, it would be great to describe, how to convert a part of `unsupported` into a module. Is it only starting a new repository in the `gismo` organisation and linking it from `gismo/CMakeLists.txt` or is there more to do?

### Option B
We can go the other way and embrace the "_one community, one library_" policy. Essentially, this would mean that everyone would be working in their branches of `stable` and once things would be ready, these would be reviewed and merged into `stable`.
#### Pros
+ Simpler and unified outline of the library.
+ One CI framework, one set of installation instructions, less CMake magic.
+ Readers of a scientific paper with G+Smo implementation might find it easier to find and test the source code (e.g., #291)
#### Cons
+ One monolithic library (is it really a problem?).
+ New algorithms instantly publicly available.

### Option C
Any ideas?

**Comment by filiatra** [comment]
I agree with your points, the optimal is probably a mix of the two.

Option B can work only if the developers have enough discipline, that is: if they just go around and touch the files in stable in their branch, nobody guarantees that these changes can be applied/accepted without breaking the rest of the library.
If there is a rule/convention that in a certain branch we are not allowed to touch files other than eg. gsElasticity if that is the module in question, this can work.
Also, I would add another disadvantage in option B: if certain users need only parts of the library they will experience high compile times and a lot of uneeded codes.

**Reply by osmihran** [reply]
> Option B can work only if the developers have enough discipline, that is: if they just go around and touch the files in stable in their branch, nobody guarantees that these changes can be applied/accepted without breaking the rest of the library.

That's why we have code review and unit tests, isn't it? I think that student A who has developed something in a quick and dirty way then has two possibilities:
- They clean it up, create a pull request and after careful review and unit tests it gets accepted;
- They don't want to invest this time and effort and the code will rot in a branch, not obstructing anything.

The advantage of code rotting in a branch (instead of rotting in devel) is that when student B comes along later and wants to build on top of the work of student A, they still have a chance of finding out how to update the rotten code: with git, they can see where the old branch diverged from stable and they can try and merge. Not ideal but usually doable.

> if certain users need only parts of the library they will experience high compile times and a lot of uneeded codes.

You are right, and it is also true that I have complained about this in the past. I think that the current compile time of entire stable library (without examples) is acceptable but it is likely to grow both in time and size if we go with option B.

**Comment by stefantakacs** [comment]
Hi,

as far as I can tell, much code in unsupported is not maintained at all.

Partly, it got superseded by newer and better written codes in stable. This happened several times, also for the multigrid code, where I had put some effort to use the new stable features also in the unsupported branch. However, this is not always possible since there might be subtle differences. Since the code from unsupported was used for the numerical experiments presented in the early papers, I never wanted to delete the old code either.

Partly, it was written by people for their PhD or master and the people have left now.

So, I think the main effort should go towards the development of stable. I myself work (almost) only with branches of stable. Some branches exist only locally, some are pushed to github.

I am not sure what to do with unsupported but I would keep it as-is, mainly as a historical record.

New development effort should go to stable (=Option B). Big separable junks of code can also go to individual new modules, for example:
* particular applications that go beyond just a few example files,
* code that depends on external software.
Key features, which include from my viewpoint at least representation of the geometry, assembling and solving, should be part of the main library.

**Reply by osmihran** [reply]
Thanks for the input, I think I mostly agree with you.

At some point you implemented sum factorization together with Andrea, haven't you? Is it using some parts of Gismo? If yes, how did you do it and what is your experience with that? Can this be taken as a suggestion of how to package things?

>  ... junks of code ...

This typo has just made my day. (=

**Reply by stefantakacs** [reply]
The sum factorization was done independently of gismo, but based on Eigen. Most of the programming was done by Andrea. We have published our code at github: https://github.com/IgASF/IgASF

I would have liked to make this available in gismo as well, so i have suggested to Angelos that this could be a gsoc project.

**Comment by osmihran** [comment, marked as answer]
Option A has been chosen and implemented in #510 .
